File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.java
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static int N, dasom, res = 0;
    static PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
//        br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_1417_국회의원선거/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        if(N == 1){
            System.out.println(0);
            return;
        }

        dasom = Integer.parseInt(br.readLine());
        for(int i = 1; i < N; i++) {
            int num = Integer.parseInt(br.readLine());
            pq.add(num);
        }

        while(!pq.isEmpty() && dasom <= pq.peek()) {
            int tmp = pq.poll();
            dasom ++;
            res ++;
            pq.add(tmp-1);
        }

        bw.write(String.valueOf(res));
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

void solve() {
    int N, dasom, res = 0;
    cin >> N;

    if (N == 1) {
        cout << "0" << "\n";
        return;
    }

    cin >> dasom;
    priority_queue<int> pq;
    for (int i = 1; i < N; i++) {
        int num;
        cin >> num;
        pq.push(num);
    }

    while (!pq.empty() && dasom <= pq.top()) {
        int tmp = pq.top();
        pq.pop();
        dasom++;
        res++;
        pq.push(tmp - 1);
    }

    cout << res << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}